src/main/java/com/atlassian/bitbucket/jenkins/internal/http/HttpRequestExecutorImpl.java

```diff
@@ -39,61 +39,38 @@ public HttpRequestExecutorImpl(Call.Factory httpCallFactory) {
-                             ResponseConsumer<T> consumer, Headers headers) {
+    public <T> T executePost(HttpUrl url, String requestBodyAsJson, ResponseConsumer<T> consumer,
+                             RequestConfiguration... additionalConfig) {
         Request.Builder requestBuilder =
-                new Request.Builder().post(RequestBody.create(JSON, requestBodyAsJson)).url(url).headers(headers);
-        return executeRequest(requestBuilder, credential, consumer);
+                new Request.Builder().post(RequestBody.create(JSON, requestBodyAsJson)).url(url);
+        return executeRequest(requestBuilder, consumer, toSet(additionalConfig));
     }
 
     @Override
-    public <T> T executePut(HttpUrl url, BitbucketCredentials credentials, String requestBodyAsJson,
-                            ResponseConsumer<T> consumer) {
+    public <T> T executePut(HttpUrl url, String requestBodyAsJson,
+                            ResponseConsumer<T> consumer, RequestConfiguration... additionalConfig) {
         Request.Builder requestBuilder =
                 new Request.Builder().put(RequestBody.create(JSON, requestBodyAsJson)).url(url);
-        return executeRequest(requestBuilder, credentials, consumer);
+        return executeRequest(requestBuilder, consumer, toSet(additionalConfig));
     }
 
-    private <T> T executeRequest(Request.Builder requestBuilder, BitbucketCredentials credentials,
-                                 ResponseConsumer<T> consumer) {
-        try {
-            addAuthentication(credentials, requestBuilder);
-            Response response = httpCallFactory.newCall(requestBuilder.build()).execute();
-            int responseCode = response.code();
-
-            try (ResponseBody body = response.body()) {
-                if (response.isSuccessful()) {
-                    log.fine("Bitbucket - call successful");
-                    return consumer.consume(response);
-                }
-                handleError(responseCode, body == null ? null : body.string());
-            }
-        } catch (ConnectException | SocketTimeoutException e) {
-            log.log(Level.FINE, "Bitbucket - Connection failed", e);
-            throw new ConnectionFailureException(e);
-        } catch (IOException e) {
-            log.log(Level.FINE, "Bitbucket - io exception", e);
-            throw new BitbucketClientException(e);
-        }
-        throw new UnhandledErrorException("Unhandled error", -1, null);
-    }
-
-    private void addAuthentication(BitbucketCredentials credential, Request.Builder requestBuilder) {
-        if (credential != ANONYMOUS_CREDENTIALS) {
-            requestBuilder.addHeader(AUTHORIZATION, credential.toHeaderValue());
-        }
+    private <T> T executeRequest(Request.Builder requestBuilder, ResponseConsumer<T> consumer,
+                                 Set<RequestConfiguration> additionalConfig) {
```

Comment:
If this takes varargs then you can to the `toSet` call here and avoid it in all the callers